private/pkg/git/gitobject/reader.go

```diff
@@ -0,0 +1,163 @@
+// Copyright 2020-2023 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package gitobject
+
+import (
+	"bufio"
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/bufbuild/buf/private/pkg/command"
+	"github.com/bufbuild/buf/private/pkg/normalpath"
+	"go.uber.org/multierr"
+)
+
+// exitTime is the amount of time we'll wait for git-cat-file(1) to exit.
+var exitTime = 5 * time.Second
+
+type reader struct {
+	rx      *bufio.Reader
+	tx      io.WriteCloser
+	process command.Process
+}
+
+func newReader(gitDirPath string, runner command.Runner) (*reader, error) {
+	gitDirPath = normalpath.Unnormalize(gitDirPath)
+	if err := validateDirPathExists(gitDirPath); err != nil {
+		return nil, err
+	}
+	gitDirPath, err := filepath.Abs(gitDirPath)
+	if err != nil {
+		return nil, err
+	}
+	// TODO: do we need to validate `gitDirPath` is not jumping context?
+	rx, stdout := io.Pipe()
+	stdin, tx := io.Pipe()
+	process, err := runner.Start(
```

Comment:
I *really* dislike that we're calling fork in a constructor, is there another way to do this?

Reply:
if we still call out to `git-cat-file`, id add an `Open` method

Reply:
but I want to know how bad `git-cat-file` is, it might be a disaster to have our own implementation of (which I expect)

Reply:
gitlab's git library also forks out to `git-cat-file` to read objects: https://gitlab.com/gitlab-org/gitaly/-/blob/v1.87.0/internal/git/catfile/batch.go#L49
go-git is so far the only library I know that can read objects directly from the fs: https://github.com/go-git/go-git/blob/master/storage/filesystem/object.go#L317C1-L327; it doesn't seem _crazy_, but I'm unfamiliar with this part of git and would want to have a lot of testing around this. It's probably the desired end-state, though.

Reply:
Looked more into `git-cat-file`, it can be a beast, especially with delta objects which incrementally rebuild objects that span multiple pack files. We can certainly go down the path of implementing an in-memory cat-file, but I don't think it's worth the time investment yet.

For now I've changed `NewObjectReader` to `OpenObjectReader`. Happy to change further, let me know!